machinelearning/tensorflow.md

```diff
@@ -0,0 +1,405 @@
+
+**Using Python wheel**
+Tensorflow official supports optimized onednn+acl backend for Graviton CPUs.
+```
+pip install tensorflow-cpu-aws
+```
+
+**Using Docker hub container**
+
+Launch a Graviton instance with an Ubuntu 20.04 AMI and install docker using the below instructions.
+```
+#install and enable docker
+sudo apt update
+sudo apt install -y docker.io
+sudo systemctl enable docker && sudo systemctl start docker && sudo usermod -a -G docker ubuntu
+sudo chmod 666 /var/run/docker.sock
+
+# pull the tensorflow docker container with onednn-acl optimizations enabled
+docker pull armswdev/tensorflow-arm-neoverse
+
+# launch the docker image
+docker run -it --rm -v /home/ubuntu/:/hostfs armswdev/tensorflow-arm-neoverse
+```
+
+**Using AWS DLAMI**
+
+To launch a Graviton instance with a DLAMI via EC2 console:
+
+1. Click on "Community AMIs"
+2. Filter AMIs by clicking Ubuntu and Arm (64-bit) check boxes
+3. Search for "Deep Learning" in search field to list the current Ubuntu based DLAMIs"
+4. Select the AMI: "Deep Learning AMI Graviton Tensorflow <version>(Ubuntu 20.04) <yyyy/mm/dd>"
+
+Here is the awscli snippet to find the latest AMI:
+```
+aws ec2 describe-images --region us-west-2 --filters  Name=architecture,Values=arm64 Name=name,Values="*Deep Learning AMI*" Name=owner-alias,Values=amazon  --query 'Images[] | sort_by(@, &CreationDate)[-1] | ImageId'
+```
+
+Here is the awscli snippet to launch the same:
+```
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
+```
+Once Graviton instance is launched with the above DLAMI, the platform is ready for building Tensorflow based ML inference applications.The DLAMI comes with the additional libraries preinstalled, eg: rust compilers, transformers etc, to enable transformers based (eg: bert) inferencing applications.
+
+
+# Runtime configurations for optimal performance
+Once the Tensorflow setup is ready, enable the below runtime configurations to achieve the best performance.
+
+```
+# The default runtime backend for tensoflow is eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+export TF_ENABLE_OENDNN_OPTS=1
+
+# Graviton3 (eg: c7g instances) supports BF16 format for ML acceleration. This can be enabled in oneDNN by setting the below environment variable
+export DNNL_DEFAULT_FPMATH_MODE=BF16
+
+# Tensorflow's inter and intra_op_parallelism_threads and the system openmp thread settings are critical for the optimal workload parallelization in a multi-threaded system.
+# set the inter and intra op thread count during the session creation
+session = Session(
```

Comment:
This is confusing as there is shell and python mixed together in here.

Reply:
I will split it into two separate blocks.